spatialpandas/geometry/base.py

```diff
@@ -0,0 +1,640 @@
+from functools import total_ordering
+from numbers import Integral
+from typing import Iterable
+
+import pyarrow as pa
+import numpy as np
+import re
+
+from pandas.api.types import is_array_like
+from pandas.core.arrays import ExtensionArray
+from pandas.core.dtypes.base import ExtensionDtype
+from pandas.core.dtypes.dtypes import register_extension_dtype
+
+from spatialpandas.geometry._algorithms import (
+    bounds_interleaved, bounds_interleaved_1d, _lexographic_lt,
+    extract_isnull_bytemap)
+
+try:
+    import shapely.geometry as sg
+except ImportError:
+    sg = None
+
+
+def _validate_nested_arrow_type(nesting_levels, pyarrow_type):
+    if pyarrow_type == pa.null():
+        return pa.null()
+
+    pyarrow_element_type = pyarrow_type
+    for i in range(nesting_levels):
+        if not isinstance(pyarrow_element_type, pa.ListType):
+            raise ValueError(
+                "Expected input data to have {} nested layer(s)".format(
+                    nesting_levels)
+            )
+        pyarrow_element_type = pyarrow_element_type.value_type
+    pyarrow_element_type = pyarrow_element_type
+    numpy_element_dtype = pyarrow_element_type.to_pandas_dtype()
+    if (numpy_element_dtype() is None
+            or numpy_element_dtype().dtype.kind not in ('i', 'u', 'f')):
+        raise ValueError(
+            "Invalid nested element type {}, expected numeric type".format(
+                pyarrow_element_type
+            ))
+    return pyarrow_element_type
+
+
+def _unwrap_geometry(a, element_dtype):
+    if np.isscalar(a) and np.isnan(a):
+        # replace top-level nana with None
+        return None
+    elif isinstance(a, Geometry):
+        return np.asarray(a.data)
+    elif sg and isinstance(a, sg.base.BaseGeometry):
+        return element_dtype._shapely_to_coordinates(a)
+    else:
+        return a
+
+
+class _ArrowBufferMixin(object):
+    @property
+    def buffer_values(self):
+        buffers = self.data.buffers()
+        return np.asarray(buffers[-1]).view(self.numpy_dtype)
```

Comment:
> It might be useful to have a method in pyarrow to get to the actual flat values of a nested list array?

Yeah, to traverse the nested array from numba we need the flat values and the multiple arrays of offsets into those values.  Something that tripped me up at first was how the zero-copy slices work. It seems that none of the buffers are modified in the slice, but the resulting `ListArray` gets an updated `offset` property value.  If there was a method to get the flat values, then there would be a question of what to return in the case of a slice like this.  If this method returned only the values selected in the slice, then the offsets arrays stored in `buffers` would not apply to it. Maybe a method to get the offsets arrays should adjust the offset values to take slicing into account.

Reply:
Regarding the zero-copy slice, I just added an `offsets` attribute to ListArray, which takes this into account: https://github.com/apache/arrow/pull/5759/files. At least, the values are still non-sliced, but the offsets are correctly sliced so they point to the correct places into the non-sliced values.